store/src/main/java/com/zimbra/cs/service/mail/CopyToFiles.java

```
@@ -138,26 +128,31 @@ private Try<CopyToFilesRequest> getRequestObject(Element request) {
    * @return try of a {@link MimePart}
    */
   private Try<MimePart> getAttachmentToCopy(CopyToFilesRequest request, ZimbraSoapContext context) {
-    // get mail message
+    // get messageId with UUID if available
+    final String[] uuidMsgId = request.getMessageId().split(":");
+    String accountUUID;
+    String msgId;
+    if (Objects.equals(2, uuidMsgId.length)) {
+      accountUUID = uuidMsgId[0];
+      msgId = uuidMsgId[1];
+    } else {
+      accountUUID = context.getAuthtokenAccountId();
+      msgId = request.getMessageId();
+    }
     final Try<Integer> messageIdTry =
-        Try.of(() -> Integer.parseInt(request.getMessageId()))
+        Try.of(() -> Integer.parseInt(msgId))
             .onFailure(ex -> mLog.debug(ex.getMessage()))
             .mapFailure(
                 Case(
```

How about using something cleaner like recoverWith instead of mapFailure:

```
.recoverWith(ex->Try.failure(ServiceException.PARSE_ERROR(MailConstants.A_MESSAGE_ID + " must be an integer.", ex)))```

---

I didn't use recoverWith because the name implies a recover from an error and returning a failure doesn't seem right, because it is like recovering with a failure.
The initial idea was to make sure no other exception aside ServiceException is thrown, as it is a managed exception, that's why the mapFailure.
However I see now it maps "Exception" which is too generic, so even mapFailure doesn't make much sense. I will check and propose another solution for this.

---

I agree recoverWith is cleaner to read in first case. I see there are many exceptions that could happen and the mapping may be hiding other errors. For example "getAttachmentToCopy" always returns "messageId must be an integer", which is wrong.

---

Decided to leave mapFailure on this, as I've added two cases: numberFormatException and Exception.
In the former case maps to a specific parse error, else internal error. + added tests on this.
Let me know!

---

In ServiceException.INTERNAL_ERROR we are providing the exception. You mean returning the exception in response?

---

I mean the "intenal error" seems too wide to understand what went wrong. My doubt is are we returning the actual message of real exception or not? having that message logged or returned will be hepful rather than just saying "internal error".

---

No, we are not returning it, because I am not sure we want to do it. I think it should be logged.
I won't say it is right or wrong, rather a common lack of design among APIs.

---

:heavy_check_mark: It should be logged.

---

The underlying error is logged at debug level before mapping each failure, on each method. I think it's better to log at ERROR level, agree?